### DISCUSSION TITLE
Is it possible to decrease the maximum refinement level in a uninterested region?

### DISCUSSION THREAD

**Discussion by Pandakii** [AMReX-Codes/amrex#4194] [Q&A]
Hi everyone,
To decrease computational cost, I want to employ a relatively lower max_level in some uninterested regions, while still using a higher max_level in regions I am interested in. I am not sure if it is possible to achieve that.

**Comment by asalmgren** [comment]
You don’t need to set a different max_level in different regions - you can
control that entirely through the refinement criteria

Ann Almgren
Senior Scientist; Dept. Head, Applied Mathematics
Pronouns: she/her/hers

**Reply by Pandakii** [reply]
I just want to fix the refinement level in a specific region, such as the exit of a channel. For example, the max_level in entire domain of a channel is set to 3 but a region near exit has a constant refinement level of 2. But generally this specific region still be refined by the max_level once the refinement criteria is reached here. I know that a specific region can be fixed at the max_level by the method of static refinement. I am not sure if it is possible to have a constant refinement level lower than the max_level in a specific region.

**Comment by asalmgren** [comment]
Everything you want to do can be accomplished by setting your refinement
criteria correctly -- for example, you will never create a finest_level of
3 if your refinement criteria in that region don't trigger refinement there.

If you write your own refinement criteria, you can use tests such as

if (x > 5 && phi > 0.1)

so that you would only use that requirement when x > 5, so this would allow
you to not refine when x <= 5.

I also find this documentation helpful:
https://amrex-fluids.github.io/IAMR/Software.html#sec-tagging

On Sat, Oct 19, 2024 at 7:05 PM Pandakii ***@***.***> wrote:

> I just want to fix the refinement level in a specific region, such as the
> exit of a channel. For example, the max_level in entire domain of a channel
> is set to 3 but a region near exit has a constant refinement level of 2.
> But generally this specific region still be refined by the max_level once
> the refinement criteria is reached here. I know that a specific region can
> be fixed at the max_level by the method of static refinement. I am not sure
> if it is possible to have a constant refinement level lower than the
> max_level in a specific region.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/discussions/4194#discussioncomment-10994008>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRE6YV7WRYIDU6GHLL4GBDZ4MFX5AVCNFSM6AAAAABQHHN4GOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTAOJZGQYDAOA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Ann Almgren
Senior Scientist; Dept. Head, Applied Mathematics
Pronouns: she/her/hers

**Reply by WeiqunZhang** [reply]
To expand on @asalmgren's answer, AmrCore has a virtual function `void ErrorEst (int lev, TagBoxArray& tags, Real time, int ngrow) override = 0;` that you override. You can write something like
```
void MyAmr::ErrorEst (int lev, TagBoxArray& tags, Real, int)
{
    for (MFIter ...)
    {
        Box const& bx = ...;
        amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE (int i, int j, int k)
        {
            if (lev == 2) {
                if (this_cell_is_in_a_region_I_dont_want_to_refine) {
                      // do nothing
                }
            }
        });
    }
}
```

**Reply by Pandakii** [reply]
Thanks@asalmgren @WeiqunZhang. I wrote below refinement parameters in my .inp file in PeleC. I found that the region R2 still can be refined by level 2, even though the max_level has been set to 1. The virtual function ErrorEst may be a  valid method to achieve that but I have no idea how to use it. I would really appreciate it if you could provide an example case.

tagging.refinement_indicators = R1 R2
tagging.R1.max_level = 2
tagging.R1.adjacent_difference_greater = 100
tagging.R1.field_name = Temp
tagging.R1.in_box_lo = 5 0 0
tagging.R1.in_box_hi = 10 10 10
tagging.R2.max_level = 1
tagging.R2.adjacent_difference_greater = 100
tagging.R2.field_name = Temp
tagging.R2.in_box_lo = 0 0 0
tagging.R2.in_box_hi = 5 10 10

**Reply by WeiqunZhang** [reply]
`tagging.refine_indicators` etc. are not AMReX's own parameters. They are implemented by applications. (Most applications do not even have these parameters.) It's very possible that the exact meaning of the these parameters might be different between IAMR and PeleC. So you might consider asking this question on PeleC's GitHub site.

As for ErrorEst, I mentioned that simply because I didn't know what code you were using. If it were a code you developed, the simplest approach would be to override that function. However, in this case, I am sure PeleC developer could help you.

Ping @drummerdoc who is a developer of AMReX and Pele. He should be able to help.

**Reply by Pandakii** [reply]
Thanks for your reply. I will seek assistance from @drummerdoc @baperry2

**Reply by drummerdoc** [reply]
Yeah, those are Pele-specific knobs.  Explicit derefining is tricky to get right.  You can use a variant of the advice @asalmgren gave above though because you can restrict the use of each criteria to subregion in space (and even in time). So, you can make several different tagging indicators based on the same criteria but over different subregions.  If this doesn't solver your issue, reach out to me directly at Marc.Day@nrel.gov. We'll figure something out.
